###### ------- utils -------#######
rdirichlet2 <- function(n, alpha) {
  vct <- colMeans(rdirichlet(n, alpha))
  vct[is.na(vct)] <- 0
  vct
}
mutate_IsoPct <- function(w, ref) {
  o <- order(ref, decreasing = T)[1:2]
  w[rev(o)] <- w[o]
  w
}
mutated_IsoPct <- function(ref) {
  o <- order(ref, decreasing = T)[1:2]
  w <- rep(0, length(ref))
  w[o] <- 1
  w
}


library(gtools)
library(dplyr)
library(vroom)
library(biomaRt)
library(this.path)
setwd(dirname(this.path()))
##### variables #####
meandisp.file <- "../files/Pickrell.Cheung.Mu.Phi.Estimates.rds"
meandisp <- readRDS(meandisp.file)
meanvect <- meandisp$pickrell.cheung.mu
dispvect <- meandisp$pickrell.cheung.phi
librarysize <- 40000000
nbr_per_group <- 4
out_dir <- "../profiles/"
ensembl <- useMart(
  "ENSEMBL_MART_ENSEMBL",
  dataset = "hsapiens_gene_ensembl",
  host = "sep2019.archive.ensembl.org"
)
select_as_events <- F

##### read the isoform result file generated by RSEM #####
isoform_results_file <- "../files/SRR493366.isoforms.mod.results"
isoform.initial <- vroom(
  isoform_results_file,
  delim = "\t", col_types = cols()
) %>% as.data.frame()

##### create a gene summary data frame from the isoform data frame #####
gene_summary <- group_by(isoform.initial, gene_id) %>%
  summarise(
    expected_gene_count_gr1 = sum(expected_count),
    effective_gene_length = sum(effective_length * IsoPct / 100),
    nbr_isoforms = length(IsoPct),
    nbr_expr_isoforms = length(which(IsoPct > 0)),
    nbr_expr_isoforms10 = length(which(IsoPct > 10))
  ) %>%
  ungroup() %>%
  as.data.frame()

##### decide whether simulate differential expressed genes or not #####
## harbor no differential expressed genes ##
gene_summary$gene_de_status <- 0
gene_summary$expected_gene_count_gr2 <- gene_summary$expected_gene_count_gr1
##### magnify gene counts to meet the library size #####
gene_summary$expected_gene_count_gr1 <- gene_summary$expected_gene_count_gr1 /
  sum(gene_summary$expected_gene_count_gr1) * librarysize
gene_summary$expected_gene_count_gr2 <- gene_summary$expected_gene_count_gr2 /
  sum(gene_summary$expected_gene_count_gr2) * librarysize

##### assign dispersion to each gene in each conditions #####
gene_summary$dispersion_gr1 <- sapply(
  gene_summary$expected_gene_count_gr1,
  function(w) dispvect[which.min(abs(meanvect - w))]
)
gene_summary$dispersion_gr2 <- sapply(
  gene_summary$expected_gene_count_gr2,
  function(w) dispvect[which.min(abs(meanvect - w))]
)

##### find out which chromosome do genes belong to #####
bm <- getBM(
  attributes = c("ensembl_gene_id_version", "chromosome_name"),
  filters = "ensembl_gene_id_version",
  values = gene_summary$gene_id, mart = ensembl
)
gene_summary <- merge(gene_summary, bm, by.x = "gene_id", by.y = "ensembl_gene_id_version", all = T)
## PAR doesn't appear in package biomaRt ##
gene_summary$chromosome_name[is.na(gene_summary$chromosome_name)] <- "PAR"

##### simulate gene counts for the desired number of samples and calculate RPK for genes #####
for (i in 1:nbr_per_group) {
  gene_summary[, paste0("s", i, "_geneCount")] <-
    sapply(1:nrow(gene_summary), function(j) {
      rnbinom(
        n = 1, size = 1 / gene_summary$dispersion_gr1[j],
        mu = gene_summary$expected_gene_count_gr1[j]
      )
    })
  gene_summary[, paste0("s", i, "_geneRPK")] <-
    gene_summary[, paste0("s", i, "_geneCount")] / gene_summary$effective_gene_length * 1e3
}

for (i in (nbr_per_group + 1):(2 * nbr_per_group)) {
  gene_summary[, paste0("s", i, "_geneCount")] <-
    sapply(1:nrow(gene_summary), function(j) {
      rnbinom(
        n = 1, size = 1 / gene_summary$dispersion_gr2[j],
        mu = gene_summary$expected_gene_count_gr2[j]
      )
    })
  gene_summary[, paste0("s", i, "_geneRPK")] <-
    gene_summary[, paste0("s", i, "_geneCount")] / gene_summary$effective_gene_length * 1e3
}

isoform_summary <- isoform.initial
##### add columns indicating the most expressed and the second most expressed isoform for each genes #####
id_major <- group_by(isoform_summary, gene_id) %>%
  mutate(id_major = transcript_id[which.max(IsoPct)]) %>%
  ungroup() %>%
  as.data.frame() %>%
  .$id_major %>%
  unique()
isoform_summary$major <- isoform_summary$transcript_id %in% id_major

id_major2 <- isoform_summary[-which(isoform_summary$major == T), ] %>%
  group_by(gene_id) %>%
  mutate(id_major2 = transcript_id[which.max(IsoPct)]) %>%
  ungroup() %>%
  as.data.frame() %>%
  .$id_major2 %>%
  unique()
isoform_summary$major2 <- isoform_summary$transcript_id %in% id_major2

##### run only once to define genes harboring alternative splicing events #####
if (select_as_events) {
  index_gr1 <- which(
    gene_summary$expected_gene_count_gr1 > 500 & gene_summary$nbr_expr_isoforms10 >= 2
  )
  asta_common_file <- "../files/asta_table_common_coor.txt.gz" # asta_common <- read.table(asta_common_file, sep = "\t", header = F)
  candi <- read.table(asta_common_file, sep = "\t", header = F) %>% dplyr::filter(
    V9 %in% gene_summary$gene_id[index_gr1]
  ) # candi <- dplyr::filter(asta_common, V9 %in% gene_summary$gene_id[index_gr1]) # candi <- asta_common[asta_common[, 9] %in% gene_summary$gene_id[index_gr1], ]
  ##### mark out common events which will be alternative splicing between two conditions if switched the relative expression percentage of two main isoforms #####
  candi$idct <- 0
  for (i in 1:nrow(candi)) {
    tmp_gene <- candi[i, 9]
    isos <- strsplit(candi[i, 10], ",") %>% unlist()
    isos_1 <- strsplit(isos[1], "/") %>% unlist()
    isos_2 <- strsplit(isos[2], "/") %>% unlist()

    iso_max <- isoform_summary$transcript_id[isoform_summary$gene_id == tmp_gene & isoform_summary$major == T]
    iso_max2 <- isoform_summary$transcript_id[isoform_summary$gene_id == tmp_gene & isoform_summary$major2 == T]

    if (iso_max %in% isos_1 && iso_max2 %in% isos_2) candi$idct[i] <- 1
    if (iso_max %in% isos_2 && iso_max2 %in% isos_1) candi$idct[i] <- 1
  }

  EE <- dplyr::filter(candi, V13 == "1-2^,3-4^", idct == 1)
  IR <- dplyr::filter(candi, V13 == "0,1^2-", idct == 1)
  AD <- dplyr::filter(candi, V13 == "1^,2^", idct == 1)
  AA <- dplyr::filter(candi, V13 == "1-,2-", idct == 1)
  ES <- dplyr::filter(candi, V13 == "0,1-2^", idct == 1)

  genes_remain <- candi[candi$idct == 1, 9] %>% unique()
  length(unique(dplyr::filter(EE, V9 %in% genes_remain)[, 9]))
  genes_ee <- dplyr::filter(EE, V9 %in% genes_remain)[, 9] %>%
    unique() %>%
    sample(44)
  ee <- dplyr::filter(EE, V9 %in% genes_ee)
  genes_remain <- setdiff(genes_remain, genes_ee)

  length(unique(dplyr::filter(IR, V9 %in% genes_remain)[, 9]))
  genes_ir <- dplyr::filter(IR, V9 %in% genes_remain)[, 9] %>%
    unique() %>%
    sample(110)
  ir <- dplyr::filter(IR, V9 %in% genes_ir)
  genes_remain <- setdiff(genes_remain, genes_ir)

  length(unique(dplyr::filter(AD, V9 %in% genes_remain)[, 9]))
  genes_ad <- dplyr::filter(AD, V9 %in% genes_remain)[, 9] %>%
    unique() %>%
    sample(282)
  ad <- dplyr::filter(AD, V9 %in% genes_ad)
  genes_remain <- setdiff(genes_remain, genes_ad)

  length(unique(dplyr::filter(AA, V9 %in% genes_remain)[, 9]))
  genes_aa <- dplyr::filter(AA, V9 %in% genes_remain)[, 9] %>%
    unique() %>%
    sample(282)
  aa <- dplyr::filter(AA, V9 %in% genes_aa)
  genes_remain <- setdiff(genes_remain, genes_aa)

  length(unique(dplyr::filter(ES, V9 %in% genes_remain)[, 9]))
  genes_es <- dplyr::filter(ES, V9 %in% genes_remain)[, 9] %>%
    unique() %>%
    sample(282)
  es <- dplyr::filter(ES, V9 %in% genes_es)

  genes_sel <- c(genes_ee, genes_ir, genes_ad, genes_aa, genes_es)
  all_aff <- dplyr::filter(candi, idct == 1, V9 %in% genes_sel)

  write.table(
    all_aff, sprintf("%s/all_affected.txt", out_dir),
    sep = "\t", quote = F, col.names = F, row.names = F
  )
} else {
  all_aff <- read.table(sprintf("%s/all_affected.txt", out_dir), sep = "\t")
}

for (nsim in 1:10) {
  mydir <- sprintf("%s/sim_%i", out_dir, nsim)
  if (!dir.exists(mydir)) dir.create(mydir)
  setwd(mydir)

  isoform_summary <- dplyr::select(
    isoform_summary,
    transcript_id,
    gene_id,
    length,
    effective_length,
    expected_count,
    TPM,
    FPKM,
    IsoPct,
    major,
    major2
  ) %>% as.data.frame()

  ##### mutate isoform expression percentage slightly for each samples #####
  for (i in 1:(2 * nbr_per_group)) {
    isoform_summary <- group_by(isoform_summary, gene_id) %>%
      mutate(IsoPctDirichlet = c(rdirichlet2(1, IsoPct / 100 * 100))) %>%
      setNames(c(colnames(isoform_summary), paste0("s", i, "_IsoPct"))) %>%
      ungroup() %>%
      as.data.frame()
  }

  ##### switch the top 2 expressed isoforms of selected differential alternative splicing genes #####
  ds_genes <- unique(all_aff[, 9])
  isoform_summary_nonds <- isoform_summary[!(isoform_summary$gene_id %in% ds_genes), ]
  isoform_summary_ds <- isoform_summary[isoform_summary$gene_id %in% ds_genes, ]

  isoform_summary_ds <- group_by(isoform_summary_ds, gene_id) %>%
    mutate(
      s5_IsoPct = mutate_IsoPct(s5_IsoPct, IsoPct),
      s6_IsoPct = mutate_IsoPct(s6_IsoPct, IsoPct),
      s7_IsoPct = mutate_IsoPct(s7_IsoPct, IsoPct),
      s8_IsoPct = mutate_IsoPct(s8_IsoPct, IsoPct),
      diff_IsoPct = -diff(sort(IsoPct, decreasing = T))[1] / 100,
      gene_ds_status = 1,
      transcript_ds_status = mutated_IsoPct(IsoPct)
    ) %>%
    ungroup() %>%
    as.data.frame()

  isoform_summary_nonds <- group_by(isoform_summary_nonds, gene_id) %>%
    mutate(
      diff_IsoPct = -diff(sort(IsoPct, decreasing = T))[1] / 100,
      gene_ds_status = 0,
      transcript_ds_status = 0
    ) %>%
    ungroup() %>%
    as.data.frame()

  isoform_summary <- rbind(isoform_summary_nonds, isoform_summary_ds)

  ##### merge the gene data frame and the isoform data frame #####
  final_summary <- merge(
    isoform_summary, gene_summary,
    by.x = "gene_id", by.y = "gene_id", all = T
  )

  ##### calculate RPK for each isoforms by dividing gene RPK according to IsoPct columns #####
  for (i in 1:(2 * nbr_per_group)) {
    final_summary[, paste0("s", i, "_isoformRPK")] <-
      final_summary[, paste0("s", i, "_geneRPK")] * final_summary[, paste0("s", i, "_IsoPct")]
  }

  ##### calculate count for isoforms #####
  for (i in 1:(2 * nbr_per_group)) {
    final_summary[, paste0("s", i, "_isoformCount")] <-
      final_summary[, paste0("s", i, "_isoformRPK")] * final_summary$effective_length / 1000
  }

  ##### calculate TPM for isoforms by zooming the original TPM according to the ratios between final simulated counts and original counts #####
  for (i in 1:(2 * nbr_per_group)) {
    final_summary[, paste0("s", i, "_isoformTPM")] <-
      round(final_summary[, paste0("s", i, "_isoformCount")] /
        final_summary$expected_count * final_summary$TPM, 2)
  }

  final_summary[is.na(final_summary)] <- 0

  ##### scale each isoformTPM column so that it sums to 1 million #####
  idx <- grep("isoformTPM", colnames(final_summary))
  for (i in idx) {
    final_summary[, i] <- final_summary[, i] / sum(final_summary[, i]) * 1e6
  }

  ##### output the final summary data frame #####
  write.table(
    final_summary, "simulation_details.txt",
    row.names = F, col.names = T, quote = F, sep = "\t"
  )

  ##### output files for each simulated samples #####
  for (i in 1:(2 * nbr_per_group)) {
    tmp <- isoform.initial
    tmp$TPM <-
      final_summary[
        match(tmp$transcript_id, final_summary$transcript_id), paste0("s", i, "_isoformTPM")
      ]
    tmp$TPM <- round(tmp$TPM, digits = 2)
    tmp$TPM <- as.character(tmp$TPM)
    tmp$TPM[tmp$TPM == "0"] <- "0.00"
    tmp$expected_count <- as.character(tmp$expected_count)
    tmp$expected_count[tmp$expected_count == "0"] <- "0.00"
    tmp$FPKM <- as.character(tmp$FPKM)
    tmp$FPKM[tmp$FPKM == "0"] <- "0.00"
    tmp$IsoPct <- as.character(tmp$IsoPct)
    tmp$IsoPct[tmp$IsoPct == "0"] <- "0.00"
    write.table(
      tmp,
      file = sprintf("sample%i.txt", i),
      row.names = F, col.names = T, quote = F, sep = "\t"
    )
  }

  ##### save the whole working environment #####
  save.image(file = "sim.RData", compress = T)
}


##### plot the simulation scenario #####
library(dplyr)
library(ggplot2)
library(scales)

##### variable #####
out_dir <- "../profiles/"

for (nsim in 1:10) {
  ##### read the simulation details file generated by codes above #####
  details <- read.delim(
    sprintf("%s/sim_%i/simulation_details.txt", out_dir, nsim),
    header = T, as.is = T
  )

  details$gr1_isoformTPM <- 1 / 4 * (
    details$s1_isoformTPM + details$s2_isoformTPM + details$s3_isoformTPM + details$s4_isoformTPM
  )
  details$gr2_isoformTPM <- 1 / 4 * (
    details$s5_isoformTPM + details$s6_isoformTPM + details$s7_isoformTPM + details$s8_isoformTPM
  )
  details$transcript_ds_cat <- ifelse(
    details$transcript_ds_status == 0, "non-differentially used", "differentially used"
  )
  x <- details %>%
    group_by(gene_id) %>%
    summarise(
      s1_geneTPM = sum(s1_isoformTPM),
      s2_geneTPM = sum(s2_isoformTPM),
      s3_geneTPM = sum(s3_isoformTPM),
      s4_geneTPM = sum(s4_isoformTPM),
      s5_geneTPM = sum(s5_isoformTPM),
      s6_geneTPM = sum(s6_isoformTPM),
      s7_geneTPM = sum(s7_isoformTPM),
      s8_geneTPM = sum(s8_isoformTPM),
      gr1_geneTPM = 1 / 4 * (
        s1_geneTPM + s2_geneTPM + s3_geneTPM + s4_geneTPM
      ),
      gr2_geneTPM = 1 / 4 * (
        s5_geneTPM + s6_geneTPM + s7_geneTPM + s8_geneTPM
      ),
      s1_geneCount = sum(s1_isoformCount),
      s2_geneCount = sum(s2_isoformCount),
      s3_geneCount = sum(s3_isoformCount),
      s4_geneCount = sum(s4_isoformCount),
      s5_geneCount = sum(s5_isoformCount),
      s6_geneCount = sum(s6_isoformCount),
      s7_geneCount = sum(s7_isoformCount),
      s8_geneCount = sum(s8_isoformCount),
      gr1_geneCount = 1 / 4 * (
        s1_geneCount + s2_geneCount + s3_geneCount + s4_geneCount
      ),
      gr2_geneCount = 1 / 4 * (
        s5_geneCount + s6_geneCount + s7_geneCount + s8_geneCount
      ),
      status_ds = gene_ds_status[1]
    )

  x$status_ds <- ifelse(
    x$status_ds == 0, "non-differentially spliced", "differentially spliced"
  )
  pdf(sprintf("%s/sim_%i/simulation_details.pdf", out_dir, nsim))
  print(
    ggplot(x, aes(x = gr1_geneCount, y = gr2_geneCount, col = status_ds)) +
      geom_abline(intercept = 0, slope = 1) +
      geom_point(alpha = 1 / 2) +
      scale_x_log10() +
      scale_y_log10() +
      scale_color_manual(values = c("red", "black"), name = "") +
      guides(colour = guide_legend(override.aes = list(size = 5))) +
      xlab("Average gene count, condition 1") +
      ylab("Average gene count, condition 2") +
      theme(
        legend.position = "bottom",
        panel.background = element_rect(fill = NA, colour = "black"),
        panel.border = element_rect(fill = NA, colour = "black", size = 1),
        panel.grid.minor.x = element_blank(),
        panel.grid.minor.y = element_blank(),
        axis.text.x = element_text(
          angle = 90, vjust = 0.5, hjust = 1,
          size = 15
        ),
        axis.text.y = element_text(size = 15),
        axis.title.x = element_text(size = 15),
        axis.title.y = element_text(size = 15),
        strip.text = element_text(size = 15),
        legend.text = element_text(size = 12),
        strip.background = element_rect(fill = NA, colour = "black")
      ) +
      ggtitle("")
  )

  print(
    ggplot(x, aes(x = gr1_geneTPM, y = gr2_geneTPM, col = status_ds)) +
      geom_abline(intercept = 0, slope = 1) +
      geom_point(alpha = 1 / 2) +
      scale_x_log10() +
      scale_y_log10() +
      scale_color_manual(values = c("red", "black"), name = "") +
      guides(colour = guide_legend(override.aes = list(size = 5))) +
      xlab("Average gene TPM, condition 1") +
      ylab("Average gene TPM, condition 2") +
      theme(
        legend.position = "bottom",
        panel.background = element_rect(fill = NA, colour = "black"),
        panel.border = element_rect(fill = NA, colour = "black", size = 1),
        panel.grid.minor.x = element_blank(),
        panel.grid.minor.y = element_blank(),
        axis.text.x = element_text(
          angle = 90, vjust = 0.5, hjust = 1,
          size = 15
        ),
        axis.text.y = element_text(size = 15),
        axis.title.x = element_text(size = 15),
        axis.title.y = element_text(size = 15),
        strip.text = element_text(size = 15),
        legend.text = element_text(size = 12),
        strip.background = element_rect(fill = NA, colour = "black")
      ) +
      ggtitle("")
  )

  print(
    ggplot(
      details, aes(x = gr1_isoformTPM, y = gr2_isoformTPM, col = transcript_ds_cat)
    ) +
      geom_abline(intercept = 0, slope = 1) +
      geom_point(alpha = 1 / 2) +
      scale_x_log10() +
      scale_y_log10() +
      scale_color_manual(values = c("red", "black"), name = "") +
      guides(colour = guide_legend(override.aes = list(size = 5))) +
      xlab("Average isoform TPM, condition 1") +
      ylab("Average isoform TPM, condition 2") +
      theme(
        legend.position = "bottom",
        panel.background = element_rect(fill = NA, colour = "black"),
        panel.border = element_rect(fill = NA, colour = "black", size = 1),
        panel.grid.minor.x = element_blank(),
        panel.grid.minor.y = element_blank(),
        axis.text.x = element_text(
          angle = 90, vjust = 0.5, hjust = 1,
          size = 15
        ),
        axis.text.y = element_text(size = 15),
        axis.title.x = element_text(size = 15),
        axis.title.y = element_text(size = 15),
        strip.text = element_text(size = 15),
        legend.text = element_text(size = 12),
        strip.background = element_rect(fill = NA, colour = "black")
      ) +
      ggtitle("")
  )
  dev.off()
}


##### generate a truth table based on simulation summary #####
library(GenomicRanges)
library(GenomicFeatures)
library(Hmisc)
library(dplyr)
library(rtracklayer)

out_dir <- "../profiles"
gtf.file <- "gencode.v32.annotation.primary_assembly.protein_coding.gtf"
nsim <- 1

tx2gene <- read.delim("tx2gene_v32.txt", header = F)
tx2gene2 <- tx2gene[match(unique(tx2gene$V2), tx2gene$V2), ] %>%
  dplyr::select(V2, V3) %>%
  setNames(c("gene_id", "gene_symbol"))

final_summary <- read.delim(
  sprintf("%s/sim_%i/simulation_details.txt", out_dir, nsim),
  as.is = T
)

gene_table <- final_summary %>%
  group_by(gene_id) %>%
  summarise(
    ds_status = unique(gene_ds_status),
    de_status = unique(gene_de_status),
    ds_de_status = paste0(ds_status, ":", de_status),
    TPM = sum(TPM),
    nbr_isoforms = unique(nbr_isoforms),
    nbr_isoforms_above5 = length(which(IsoPct >= 5)),
    nbr_isoforms_above10 = length(which(IsoPct >= 10)),
    nbr_isoforms_above15 = length(which(IsoPct >= 15)),
    nbr_isoforms_above25 = length(which(IsoPct >= 25)),
    diff_IsoPct = unique(diff_IsoPct)
  ) %>%
  ungroup() %>%
  as.data.frame()

gene_table$TPM_2 <- cut2(gene_table$TPM, g = 2)
gene_table$TPM_5 <- cut2(gene_table$TPM, g = 5)
gene_table$TPM_10 <- cut2(gene_table$TPM, g = 10)
gene_table$nbr_isoforms_2 <- cut2(gene_table$nbr_isoforms, g = 2)
gene_table$nbr_isoforms_5 <- cut2(gene_table$nbr_isoforms, g = 5)
gene_table$nbr_isoforms_10 <- cut2(gene_table$nbr_isoforms, g = 10)
gene_table$diff_IsoPct_2 <- cut2(gene_table$diff_IsoPct, g = 2)
gene_table$diff_IsoPct_5 <- cut2(gene_table$diff_IsoPct, g = 5)
gene_table$diff_IsoPct_10 <- cut2(gene_table$diff_IsoPct, g = 10)
gene_table$diff_IsoPct_3eq <- cut2(gene_table$diff_IsoPct,
  cuts = c(0, 1 / 3, 2 / 3, 1)
)

tmp <- final_summary %>%
  group_by(gene_id) %>%
  summarise(
    topisoform = transcript_id[which.max(IsoPct)],
    secisoform = ifelse(
      length(transcript_id) > 1,
      transcript_id[order(IsoPct, decreasing = T)[2]],
      "NA"
    )
  ) %>%
  ungroup() %>%
  as.data.frame()
txdb <- makeTxDbFromGFF(gtf.file, format = "gtf")
ebt <- exonsBy(txdb, "tx", use.names = T)
ebt2 <- ebt[setdiff(c(tmp$topisoform, tmp$secisoform), "NA")]
tmp$diffbp <- sapply(1:nrow(tmp), function(i) {
  if (any(tmp[i, ] == "NA")) {
    0
  } else {
    a <- ebt2[[unlist(tmp[i, "topisoform"])]]
    b <- ebt2[[unlist(tmp[i, "secisoform"])]]
    sum(width(GenomicRanges::union(a, b))) - sum(width(GenomicRanges::intersect(a, b)))
  }
})
tmp$diffbp_2 <- cut2(tmp$diffbp, g = 2)
tmp$diffbp_5 <- cut2(tmp$diffbp, g = 5)
tmp$diffbp_10 <- cut2(tmp$diffbp, g = 10)
gene_table <- merge(
  gene_table,
  tmp[, c("gene_id", "diffbp", "diffbp_2", "diffbp_5", "diffbp_10")],
  by = "gene_id",
  all = T
)

# idx <- which(colnames(gene_table) == "gene_id")
# colnames(gene_table)[idx] <- "gene"

gene_table <- dplyr::left_join(gene_table, tx2gene2)
gene_table <- gene_table %>%
  dplyr::select(gene_id, gene_symbol, dplyr::everything())

write.table(
  gene_table, file = sprintf("%s/sim_%i/truth.txt", out_dir, nsim),
  row.names = F, col.names = T, sep = "\t", quote = F
)




